Fold map function for List

bschommer opened the pull request:
As the title indicates this PR implements the quite common and popular function that combines `fold` and `map`. I don't have any strong opinion on the order of arguments and or the name.

dra27 commented:
A few comments, not including an opinion on whether this *should* go in:

- Batteries has `BatList.fold_left_map` (see the discussion when it was added in https://github.com/ocaml-batteries-team/batteries-included/pull/734).
- If merged, `ListLabels` needs updating too.
- Order of arguments should match `fold_left` (which it does - that's a case for having a strong opinion!)
- Containers also has the function, same signature, but called `fold_map`
- Base also has the function, with two variants `folding_map` (which doesn't return the accumulator) and `fold_map` (which does)

bschommer commented:
I took inspiration from the above mentioned implementation, the name is inspired by the recently added `filter_map`. `ListLabels` is also updated. As said I don't have a strong opinion on the name or argument order but I would like to have this function included since it is rather common and I implemented it myself often enough to wish it to be in the standard library.

bschommer commented:
The failing CI build is due to missing changes, I wanted to wait until the name is fixed before adding them.

cfcs commented on testsuite/tests/lib-list/test.ml:
```diff
@@ -42,6 +42,7 @@ let () =
   assert (List.compare_length_with ['1'] 1 = 0);
   assert (List.compare_length_with ['1'] 2 < 0);
   assert (List.filter_map string_of_even_opt l = ["0";"2";"4";"6";"8"]);
+  assert (List.fold_map (fun a b -> a + b, b) 0 l = (45, l));
```
nit, but this test case doesn't check that the `map` part of `fold_map` does anything.

bschommer replied:
I added some more tests.

cfcs commented:
Err, sorry to be back with more annoying bits :) 
1) A lot of the other functions document (in their mli docstrings) whether or not they are tail-recursive so people can avoid using them if they have very long lists.
Would it make sense to add that for this function too (since the call to `List.rev` makes it non-tail-recursive)?

2) Would there be a case for a `fold_map_rev` which was tail-recursive?

bschommer commented:
> 1. A lot of the other functions document (in their mli docstrings) whether or not they are tail-recursive so people can avoid using them if they have very long lists.
>    Would it make sense to add that for this function too (since the call to `List.rev` makes it non-tail-recursive)?

`List.rev` is actually tail-recursive.

dra27 commented:
Given that we have `rev_map` it might be nice to have the *faster* `rev_fold_map` as well, but that doesn't have to be done here.

cfcs commented:
> `List.rev` is actually tail-recursive.

Woops, my bad. Indeed, but it's a bit slower as you noted. :)

bschommer commented:
> Given that we have `rev_map` it might be nice to have the _faster_ `rev_fold_map` as well, but that doesn't have to be done here.

I think if we can agree on the name and arguments of the function, we can actually do even a bit more and add the function to all modules that have a `map` function.

bschommer commented:
I rebased the branch. I still think this is a useful and often used function and would like to have it included in the standard library.

alainfrisch commented:
Thanks for the contribution.

I agree this is a useful addition. I have a preference for `fold_left_map`, which is more coherent with the rest of the Stdlib.  Even in `Seq` where the traversal direction is unambiguous, we have `fold_left`, not just `fold`.

Does anyone prefer to keep `fold_map` or push for something else?

>  we can actually do even a bit more and add the function to all modules that have a map function.

Indeed, one can even generically transform a `map` function to an accumulating variant:
````ocaml
let map_accu map f acc l =
  let acc = ref acc in
  let g x =
    let new_acc, y = f !acc x in
    acc := new_acc;
    y
  in
  let r = map g l in
  !acc, r
````

and this shows that providing the function is "just" a convenience for common cases.  But let's keep that for later. I find it much less natural for data structures that implements a `map` but which don't have a canonical traversal ordering.  E.g. `Map` is borderline (its `map` function specifies the traversal ordering, so that would be ok-ish), but if we had an `Hashtbl.map` (copying an hashtable, mapping values, with an unspecified traversal ordering), that wouldn't be so good.

lpw25 commented:
+1 to naming it `fold_left_map`. To me `fold` means `fold_right`.

bschommer commented:
I'm okay with `fold_left_map` the name then also makes the iteration order more clear.

alainfrisch commented on stdlib/listLabels.mli:
```diff
@@ -159,6 +159,13 @@ val concat_map : f:('a -> 'b list) -> 'a list -> 'b list
+  f:('a -> 'b -> 'a * 'c) -> init:'a -> 'b list -> 'a * 'c list
+(** [fold_left_map] is  a combination of [fold_left] and [map] hat threads an
+    accumulator through calls to [f]
+    @since 4.10.0
```
```suggestion
    @since 4.11.0
```

alainfrisch commented on stdlib/list.mli:
```diff
@@ -148,6 +148,12 @@ val concat_map : ('a -> 'b list) -> 'a list -> 'b list
+val fold_left_map : ('a -> 'b -> 'a * 'c) -> 'a -> 'b list -> 'a * 'c list
+(** [fold_left_map] is  a combination of [fold_left] and [map] that threads an
+    accumulator through calls to [f]
+    @since 4.10.0
```
```suggestion
    @since 4.11.0
```

alainfrisch commented on Changes:
```diff
@@ -56,6 +56,11 @@ OCaml 4.10.0
 
+### Standard library:
+
+- #8894: Added List.fold_left_map function combining map and fold.
```
To be moved to current working version (=> 4.11).

alainfrisch commented:
LGTM. Could you move the change entry to 4.11 and update the since tags accordingly? Then it should be good to merge if nobody complains.

bschommer committed:
Added fold_left_map function.

The fold_map function is quite common combination of fold and map
which allow it to pass an accumulator through map.

bschommer committed:
Added test for the new function

bschommer committed:
Added Change entry.

bschommer commented:
I cleaned up the history, changed the target version and add @cfcs and @alainfrisch as reviewer.

alainfrisch reviewed (approved):
LGTM, will merge soon if nobody complains.

@bschommer : thanks for your contribution (in general, no need to put efforts rewriting the history since we tend to squash-merge small PRs anyway).

bschommer commented:
> @bschommer : thanks for your contribution (in general, no need to put efforts rewriting the history since we tend to squash-merge small PRs anyway).

No problem.
